model/src/pyrenew/process/rtperiodicdiff.py

```diff
@@ -0,0 +1,277 @@
+
+        return jnp.hstack([log_rt1, next_log_rt, b]), next_log_rt
+
+    def sample(
+        self,
+        duration: int | None = None,
+        **kwargs,
+    ) -> RtPeriodicDiffProcessProcessSample:
+        """
+        Samples the periodic Rt with autoregressive difference.
+
+        Parameters
+        ----------
+        duration : int, optional
+            Duration of the sequence. Defaults to None, in which case it is
+            set to the number of observations (`self.n_obs`).
+        **kwargs : dict, optional
+            Additional keyword arguments passed through to internal sample()
+            calls, should there be any.
+
+        Returns
+        -------
+        RtPeriodicDiffProcessProcessSample
+            Named tuple with "rt".
+        """
+
+        # Initial sample
+        log_rt_prior = self.log_rt_prior.sample(**kwargs)[0]
+        b = self.autoreg.sample(**kwargs)[0]
+        s_r = self.periodic_diff_sd.sample(**kwargs)[0]
+
+        # Running the process
+        ar_diff = FirstDifferenceARProcess(autoreg=b, noise_sd=s_r)
+        log_rt = ar_diff.sample(
+            duration=self.n_periods,
+            init_val=log_rt_prior[1],
+            init_rate_of_change=log_rt_prior[1] - log_rt_prior[0],
+        )[0]
+
+        # Expanding according to the number of days
+        if duration is None:
+            duration = self.n_timepoints
+        elif duration > self.n_timepoints:
+            raise ValueError(
+                "Duration should be less than or equal "
+                f"to n_timepoints ({self.n_timepoints })."
+            )
+
+        return RtPeriodicDiffProcessProcessSample(
+            rt=jnp.repeat(jnp.exp(log_rt.flatten()), self.period_size)[
+                self.data_starts : (self.data_starts + duration)
+            ],
+        )
+
+
+class RtWeeklyDiffProcess(RtPeriodicDiffProcess):
+    """
+    Weekly Rt with autoregressive difference.
+    """
+
+    def __init__(
+        self,
+        n_timepoints: int,
+        data_starts: int,
+        log_rt_prior: RandomVariable,
+        autoreg: RandomVariable,
+        periodic_diff_sd: RandomVariable,
+        site_name: str = "rt_weekly_diff",
+    ) -> None:
+        """
+        Default constructor for RtWeeklyDiffProcess class.
+
+        Parameters
+        ----------
+        n_timepoints : int
+            Size of the returned sample.
+        data_starts : int
+            Relative point at which data starts, must be between 0 and 6.
+        log_rt_prior : RandomVariable
+            Log Rt prior for the first two observations.
+        autoreg : RandomVariable
+            Autoregressive parameter.
+        periodic_diff_sd : RandomVariable
+            Standard deviation of the noise.
+        site_name : str, optional
+            Name of the site. Defaults to "rt_weekly_diff".
+
+        Returns
+        -------
+        None
+        """
+
+        assert 0 <= data_starts <= 6, (
```

Comment:
This could be a generic check for `data_starts <= period_size - 1`, which could be handled in `RtPeriodicDiffProcess`.

Reply:
Not sure I understand your question. Perhaps we can address during coworking time today.

Reply:
To address this question in a more concrete way, Do we want `assert` statements and other non-test checks as soon as the variables are _seen_ or only right before they are used.

Reply:
Addressed in 4e4a12e and 963baed.